File: src/test/resources/standardization/scipy/scripts/ResultsHandler.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

# Copyright (C) 2017-2018 Project-ODE
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

# Authors: Dorian Cazau, Alexandre Degurse

import os
import pandas
import numpy as np

from SoundHandler import SoundHandler


class ResultsHandler:
    def __init__(
        self,
        soundHandler,
        algorithm,
        nfft,
        winSize,
        offset,
        vSysBits=64
    ):

        self.soundHandler = soundHandler
        self.algorithm = algorithm
        self.nfft = nfft
        self.winSize = winSize
        self.offset = offset
        self.vSysBits = vSysBits

        self.value = None

        self.fileName = "_".join(
            [str(p) for p in [soundHandler, algorithm, nfft,
                              winSize, offset, vSysBits]
             ]) + ".csv"

    def __str__(self):
        return self.fileName

    def setValue(self, value):
        self.value = value

    def formatComplexResults(self):
        """
        Results containing complex values are reformatted following
        the same convention as in FeatureEngine, ie:
        [z_0, z_1, ... , z_n] => [Re(z_0), Im(z_0), Re(z_1), ... Im(z_n)]
        """
        initialShape = self.value.shape

        nSeg = initialShape[1]
        segLength = initialShape[0]

        valueAsScalaFormat = np.zeros((nSeg, 2*segLength), dtype=float)
        valueAsComplex = self.value.transpose()

        for i in range(nSeg):
            valueAsScalaFormat[i, ::2] = valueAsComplex[i].real
            valueAsScalaFormat[i, 1::2] = valueAsComplex[i].imag

        self.value = valueAsScalaFormat.transpose()


    def write(self):
        if self.value is None:
            raise("No values to write")

        if self.algorithm is "vFFT":
            self.formatComplexResults()

        valueDataFrame = pandas.DataFrame(self.value)

        # store using one line per time-result
        valueDataFrame = valueDataFrame.transpose()

        valueDataFrame.to_csv(os.path.join("../values/" + str(self)),
                              index=False, header=False,
                              sep=' ', float_format='%.16f')


if __name__ == "__main__":
    s = SoundHandler("Sound1", 64, 24, 9811, 3906.0, 1)
    resHandler = ResultsHandler(soundHandler=s, algorithm="vPSD", nfft=128,
                                winSize=128, offset=128)

    print(resHandler)

File: src/test/scala/org/oceandataexplorer/engine/standardization/ResultsHandler.scala
/** Copyright (C) 2017-2018 Project-ODE
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.oceandataexplorer.engine.standardization

import java.io.{File, FileInputStream, InputStream}
import scala.io.Source

import org.oceandataexplorer.engine.signalprocessing._
import org.oceandataexplorer.engine.signalprocessing.windowfunctions._
import WindowFunctionTypes.{Symmetric, Periodic}

import org.scalatest.{FlatSpec, Matchers}

/**
 * Class used to read result files with the filename format describe in README.md
 *
 * @author Alexandre Degurse
 */

case class ResultsHandler (
  sound: SoundHandler,
  algo: String,
  nfft: Int,
  winSize: Int,
  offset: Int,
  vSysBits: Int,
  location: String
) {

  val paramsString = (sound.soundParametersString
    + "_" + algo + "_" + nfft.toString + "_" + winSize.toString
    + "_" + offset.toString + "_" + vSysBits.toString
  )

  val fileName = paramsString + ".csv"
  private val file = new File(getClass.getResource(location + "/" + fileName).toURI)

  // for now, offset is specified by the file names
  private val overlap = winSize - offset

  // instanciate the signalprocessing classes
  private val segmentation = Segmentation(winSize, overlap)

  private val hammingSymmetric = HammingWindowFunction(winSize, Symmetric)

  private val hammingPeriodic = HammingWindowFunction(winSize, Periodic)
  private val hammingDensityPeriodicNormFactor = hammingPeriodic.densityNormalizationFactor()
  private val hammingSpectrumPeriodicNormFactor = hammingPeriodic.spectrumNormalizationFactor()

  private val fftClass = FFT(nfft, sound.samplingRate)
  private val periodogramClassNormInDensity = Periodogram(
    nfft,
    1 / (sound.samplingRate.toDouble * hammingDensityPeriodicNormFactor),
    sound.samplingRate
  )
  private val welchClass = WelchSpectralDensity(nfft, sound.samplingRate)

  /**
   * We're using a convention for TOL, the study frequency range is conventionally [0.2 * samplingRate, 0.4 * samplingRate]
   */
  private val tolClass = if (winSize >= sound.samplingRate) {
    TOL(nfft, sound.samplingRate, Some(0.2 * sound.samplingRate.toDouble), Some(0.4 * sound.samplingRate.toDouble))
  } else null

  /**
   * Function computing results given the class parameters
   * @return The results of the computation defined by the class parameters as a Array of Array[Double]
   */
  def getComputedValues: Array[Array[Double]] = {
    val signal = sound.readSound()

    val ffts = segmentation.compute(signal)
      .map(segment => hammingPeriodic.applyToSignal(segment))
      .map(win => fftClass.compute(win))

    algo match {
      case "vFFT" =>
        // Scipy Short Time Fourier Transform normalizes the ffts using the
        // sqrt of the spectrum normalization factor

        ffts.map(fft => fft.map(_ / math.sqrt(hammingSpectrumPeriodicNormFactor)))

      case "vPSD" =>
        ffts
          .map(fft => periodogramClassNormInDensity.compute(fft))

      case "vWelch" =>
        val periodograms = ffts
          .map(fft => periodogramClassNormInDensity.compute(fft))

        Array(welchClass.compute(periodograms))

      case "vTOL" =>
        val welch = welchClass.compute(
          ffts.map(fft => periodogramClassNormInDensity.compute(fft))
        )

        Array(tolClass.compute(welch))

      case "fWelch" => Array(welchClass.frequencyVector)
    }
  }

  /**
   * Function used to read the result file described by the class parameters
   * @return The expected values for the class parameters as a Array of Array[Double]
   */
  def getExpectedValues: Array[Array[Double]] = ResultsHandler.readResultFile(file)
}

object ResultsHandler {
  /**
   * Function that reads a file and extract its values as Double.
   * The first dimension corresponds the lines in the file.
   * The second dimension corresponds the content of a line.
   *
   * @param resultFile The file to be read as a java.io.File
   * @return The values contained in the file as a Array of Array[Double] order as described above.
   */
  def readResultFile(resultFile: File): Array[Array[Double]] = {
    val resultFileInputStream: InputStream = new FileInputStream(resultFile)
    val resultString = Source.fromInputStream(resultFileInputStream).mkString.trim.split("\n")

    resultString
      .map{s =>
        s.trim
          .split(" ")
          .map(x => x.toDouble)
      }
  }
}
